### PR TITLE
layout: Unify size calculation of replaced elements

### PR DESCRIPTION
**Pull request by stshine**
---

- [ ] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/14490)

### CODE REVIEW

**Review by asajeffrey — COMMENTED** [review]
You probably need a reviewer who knows the layout system better than me!

File: components/layout/model.rs [hunk 1]
```diff
@@ -436,6 +436,21 @@ impl MaybeAuto {
     }
 }
 
+// Return used value for inline-size or block-size.
```
Comment (emilio) [inline components/layout/model.rs]: Can you make this a doc comment? Also, this doesn't return the used value (the used value is technically the _final_ pixel value).

Probably you can reword it somehow. This returns a middle-ground between the computed value and the used value.

File: components/layout/fragment.rs [hunk 2]
```diff
@@ -1203,6 +924,181 @@ impl Fragment {
...
+            // https://drafts.csswg.org/css2/visudet.html#min-max-widths
+            (MaybeAuto::Auto, MaybeAuto::Auto) => {
+                if self.has_intrinsic_ratio() {
+                    // This approch follows the spirit of cover and contain constraint.
+                    // https://drafts.csswg.org/css-images-3/#cover-contain
+
+                    // First, have two rectangles that keep aspect ratio while may be clamped
+                    // by the contraints;
+                    let (first_isize, first_bsize) = {
+                        let isize = inline_constraint.clamp(intrinsic_inline_size);
```
Comment (emilio) [inline components/layout/fragment.rs]: Can you expand the name of `isize` so it isn't the same as a type? `inline_size` is fine, probably doing the same for `block_size`.

File: components/layout/fragment.rs [hunk 3]
```diff
@@ -1203,6 +924,181 @@ impl Fragment {
...
+                            (first_isize, first_bsize),
+                        // It does not matter which we choose here, because both sizes
+                        // will be clamped to constraint;
+                        (Ordering::Less, Ordering::Greater) | (Ordering::Greater, Ordering::Less) =>
+                            (first_isize, first_bsize)
+                    };
+                    // Clamp the result and we are done :-)
+                    (inline_constraint.clamp(inline_size), block_constraint.clamp(block_size))
+                } else {
+                    (inline_constraint.clamp(intrinsic_inline_size),
```
Comment (emilio) [inline components/layout/fragment.rs]: I guess we can early return and deindent the rest? I think that'd help to keep it a bit more legible. Not a big deal though.

File: components/layout/fragment.rs [hunk 4]
```diff
@@ -1203,6 +924,181 @@ impl Fragment {
...
+        }
+    }
+
+    /// Whether this replace element has intrinsic aspect ratio.
+    pub fn has_intrinsic_ratio(&self) -> bool {
+        match self.specific {
+            SpecificFragmentInfo::Image(_)  |
+            SpecificFragmentInfo::Canvas(_) |
+            // TODO(stshine): According to the SVG spec, whether a SVG element has intrinsic
+            // aspect ratio is determined by the `preservAspectRatio` attribute. Since for now
```
Comment (emilio) [inline components/layout/fragment.rs]: nit: `preserveAspectRatio`.

File: components/layout/model.rs [hunk 5]
```diff
@@ -507,61 +522,59 @@ impl ToGfxMatrix for ComputedMatrix {
...
 /// A min or max constraint
-///
-/// A `max` of `None` is equivalent to no limmit for the size in the given
-/// dimension. The `min` is >= 0, as negative values are illegal and by
-/// default `min` is 0.
 #[derive(Debug)]
-pub struct MinMaxConstraint {
-    min: Au,
-    max: Option<Au>
+pub struct SizeConstraint {
```
Comment (emilio) [inline components/layout/model.rs]: (It'd be nice to keep the renaming in a different commit, but that doesn't matter now :P)

File: components/layout/model.rs [hunk 6]
```diff
@@ -507,61 +522,59 @@ impl ToGfxMatrix for ComputedMatrix {
...
+        } else {
+            SizeConstraint {
+                min_size: min_size,
+                max_size: max_size
+            }
         }
     }
 
     /// Clamp the given size by the given `min` and `max` constraints.
+    /// Make sure the size is to be used for content box.
```
Comment (emilio) [inline components/layout/model.rs]: Can you elaborate on this comment? Technically this can be reused for other purposes.

File: components/layout/fragment.rs [hunk 7]
```diff
@@ -1549,66 +1444,39 @@ impl Fragment {
                 let block_flow = info.flow_ref.as_block();
                 result.union_block(&block_flow.base.intrinsic_inline_sizes)
             }
-            SpecificFragmentInfo::Image(ref mut image_fragment_info) => {
-                let mut image_inline_size = match self.style.content_inline_size() {
+            SpecificFragmentInfo::Image(_) | SpecificFragmentInfo::Canvas(_) |
```
Comment (emilio) [inline components/layout/fragment.rs]: Can we move `is_replaced_content` to `Fragment` so we can reuse it here and everywhere else like:

```rust
_ if self.is_replaced_content() => {
   // ...
}
```

or alternatively:

```rust
if self.is_replaced_content() {
    // ....
    return ...;
}

match self.specific {
     SpecificFragmentInfo::Iframe(..) | Specifi... =>
         unreachable!("Replaced content should've been handled!"),
}
```

Since I guess we care about not forgetting to add fragment types to some parts of the layout code, and a compiler error is faster and better than a runtime error.

**Review by emilio — COMMENTED** [review]
This looks great! I love negative diffs :)

Some comments below.

File: components/layout/block.rs [hunk 8]
```diff
@@ -670,16 +670,16 @@ impl BlockFlow {
...
     /// replaced fragments.
     fn is_replaced_content(&self) -> bool {
         match self.fragment.specific {
-            SpecificFragmentInfo::ScannedText(_) |
             SpecificFragmentInfo::Svg(_) |
             SpecificFragmentInfo::Image(_) |
             SpecificFragmentInfo::Canvas(_) |
-            SpecificFragmentInfo::InlineBlock(_) => true,
+            SpecificFragmentInfo::Iframe(_) => true,
+            SpecificFragmentInfo::ScannedText(_) => panic!("A block that is also a text fragment!"),
```
Comment (emilio) [inline components/layout/block.rs]: @asajeffrey This is never supposed to happen, but I agree this panic is out of place.

Comment (stshine) [inline components/layout/block.rs]: I thought I could complete this PR last night, but unfortunately ran into another bug: on replaced elements margins in block direction do not take effect. While this bug is really weird,  I will try to fix it this morning.

File: components/layout/model.rs (hunk 1 above)
Comment (stshine) [inline components/layout/model.rs]: Sure, will do, this is from removed code.

File: components/layout/model.rs (hunk 6 above)
Comment (stshine) [inline components/layout/model.rs]: Will fix in a seperate commit.

File: components/layout/fragment.rs (hunk 2 above)
Comment (stshine) [inline components/layout/fragment.rs]: There are already too many `inline_size`s, but I will found a better name.

File: components/layout/fragment.rs (hunk 3 above)
Comment (stshine) [inline components/layout/fragment.rs]: Hmm... I am not sure about this, use two branch rather than an early return feels a clearer logic.

File: components/layout/fragment.rs (hunk 7 above)
Comment (stshine) [inline components/layout/fragment.rs]: Good idea, will do. There are so many repeated `is_replaced_or_*` code in `fragment.rs`.

**Comment by stshine** [comment]
Finally this is done, I'm so happy :D (Hope I will never need to deal with margin collapse and line metric at the same time again, even the simplest hacks.)

**Comment by emilio** [comment]
Awesome, the new passes are exciting! Please update the tests expectations and tell me when you feel it's ready so I review it more in depth :)

**Comment by stshine** [comment]
This pull request is ready for review.
And, hope you will like it :-)

File: components/layout/fragment.rs [hunk 9]
```diff
@@ -1202,6 +923,179 @@ impl Fragment {
...
+                        (Ordering::Equal, Ordering::Equal) =>
+                            (first_isize, first_bsize),
+                        // When only one rectangle is clamped, use it;
+                        (Ordering::Equal, _) =>
+                            (second_isize, second_bsize),
+                        (_, Ordering::Equal) =>
+                            (first_isize, first_bsize),
+                        // When both rectangles grow (smaller than min sizes),
+                        // Choose the larger one;
+                        (Ordering::Greater, Ordering::Greater) if first_isize > second_isize =>
```
Comment (emilio) [inline components/layout/fragment.rs]: Probably this can be a bit clearer like:

```rust
(Ordering::Greater, Ordering::Greater) => {
    if first_isize > second_isize {
         // ...
    } else {
        // ...
    }
}
```

Instead of two different cases.

File: components/layout/fragment.rs [hunk 10]
```diff
@@ -1202,6 +923,179 @@ impl Fragment {
...
+
+        let inline_size = style_length(self.style.content_inline_size(), containing_inline_size)
+                                     .map(|x| x - self.box_sizing_boundary(Direction::Inline));
+        let block_size = style_length(self.style.content_block_size(), containing_block_size)
+                                     .map(|x| x - self.box_sizing_boundary(Direction::Block));
+        let inline_constraint = self.size_constraint(containing_inline_size, Direction::Inline);
+        let block_constraint = self.size_constraint(containing_block_size, Direction::Block);
+
+        // https://drafts.csswg.org/css-images-3/#default-sizing
+        match (inline_size, block_size) {
```
Comment (emilio) [inline components/layout/fragment.rs]: Spec texts change, so in general more spec quoting on top of linking I think would be helpful.

This is mostly a personal preference, and I know this algorithm in particular isn't going to change, so I won't enforce it, feel free to add quotes when you feel it's appropriate.

File: components/layout/fragment.rs (hunk 3 above)
Comment (emilio) [inline components/layout/fragment.rs]: I was mostly concerned about rightward drift, but it's ok.

File: components/layout/fragment.rs [hunk 11]
```diff
@@ -1548,66 +1441,41 @@ impl Fragment {
...
+                        self.border_padding.inline_end = model::specified(padding.inline_end, Au(0));
+                        self.border_padding.block_start = model::specified(padding.block_start, Au(0));
+                        self.border_padding.block_end = model::specified(padding.block_end, Au(0));
+                        let border = self.border_width();
+                        self.border_padding.inline_start += border.inline_start;
+                        self.border_padding.inline_end += border.inline_end;
+                        self.border_padding.block_start += border.block_start;
+                        self.border_padding.block_end += border.block_end;
+                        let (isize, _) = self.calculate_replaced_sizes(None, None);
+                        isize
```
Comment (emilio) [inline components/layout/fragment.rs]: So I understand why you're doing this, but I think that it feels weird to have two places where we initialize border_padding. @pcwalton any thoughts?

**Review by emilio — COMMENTED** [review]
The rest looks great to me, this is a nice cleanup! :)

File: components/layout/fragment.rs (hunk 11 above)
Comment (stshine) [inline components/layout/fragment.rs]: Yeah, this is done this way because 1. the `box_sizing_boundary()` gets the border info from the `border_padding` field, 2. the `surrounding_intrinsic_inline_size()` method does not return the block direction borders, 3. it is used during other layout passes other than intrinsic inline size bubbling so we can not simply store the border value in the `border_padding` field. If there is a good solution I'd love to take.

Comment (stshine) [inline components/layout/fragment.rs]: Other the other hand, my next plan is to make an enhancement to our intrinsic inline size bubbling pass. Maybe I can attack this trouble together.

File: components/layout/fragment.rs (hunk 10 above)
Comment (stshine) [inline components/layout/fragment.rs]: Added some spec texts.

**Comment by stshine** [comment]
Updated according to review comments.

File: components/layout/fragment.rs (hunk 11 above)
Comment (emilio) [inline components/layout/fragment.rs]: Can you add a TODO there?

**Review by emilio — COMMENTED** [review]
Looks good to me, modulo that, though I think it's fine to land it and then rework how it works, over all if it's already on your radar.

r=me with that TODO/FIXME comment.

File: components/layout/fragment.rs (hunk 11 above)
Comment (stshine) [inline components/layout/fragment.rs]: Sure, will do

**Comment by stshine** [comment]
@bors-servo r+

@emilio Thanks :-)

**Comment by stshine** [comment]
Oh no.
@bors-servo r-

**Comment by stshine** [comment]
@bors-servo retry

infra
